### PR TITLE
Add Umbrella Sampling

### PR DESCRIPTION
**Pull request by InnocentBug**
Implement umbrella sampling (of a single point).
This can be a building block for the string method and other methods.

### CODE REVIEW

**Comment by InnocentBug** [comment]
I just added examples and tests for umbrella sampling.

My desktop setup doesn't like to cooperate with cupy and I don't have the time to make it work rn.
Maybe I wait until I can test on midway3 with hoomd-blue.

File: pysages/ssages/methods/umbrella.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
+# SPDX-License-Identifier: GPL-3.0-or-later
```
Comment (pabloferz) [inline pysages/ssages/methods/umbrella.py:1]: Is this based on another implementation? If not I've been trying to have anything that is not derived from implementations in SSAGES MIT licensed.

Comment (InnocentBug) [reply pysages/ssages/methods/umbrella.py:1]: No, it's not based on anything.
I did pay much attention to the license.
I am fine with any

File: pysages/ssages/methods/umbrella.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    indices = helpers.indices
+
+    def initialize():
+        bias = np.zeros((natoms, 3))
+        return UmbrellaState(bias, None)
+
+    def update(state, rs, vms, ids):
+        xi, Jxi = cv(rs, indices(ids))
+        D = kspring * (xi - center)
+        bias = -D*Jxi
```
Comment (pabloferz) [inline pysages/ssages/methods/umbrella.py:46]:
```suggestion
        bias = (-D * Jxi).reshape(state.bias.shape)
```

Comment (pabloferz) [reply pysages/ssages/methods/umbrella.py:46]: Just to make sure that bias preserves its shape.

File: examples/umbrella_sampling/umbrella.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        qr += ['temperature', 'potential_energy', 'kinetic_energy']
+
+        cvs = (Component([0], 0), Component([0], 1))
+
+        method = UmbrellaSampling(cvs, 10., (-2, 2))
+
+        hoomd.md.integrate.nvt(group=hoomd.group.all(), kT=1, tau=0.1)
+        hoomd.md.integrate.mode_standard(dt=0.05)
+
+        bind(context, method)
```
Comment (pabloferz) [inline examples/umbrella_sampling/umbrella.py:31]:
```suggestion
        sampler = bind(context, method)
```

To do manual logging for tests purposes, you might want to read `sampler.state.xi` after a number of time steps. Not really necessary for the PR though.

Comment (InnocentBug) [reply examples/umbrella_sampling/umbrella.py:42]: well, with the merged #45 I decided to purposefully not assign the result of bind in order to test the global sampler pool.

**Comment by InnocentBug** [comment]
[hist.pdf](https://github.com/SSAGESLabs/PySAGES/files/7262741/hist.pdf)
I am happy, it seems to work at least somehow.

You can bias the position of a single particle and it fluctuates around a given position.
There are two things, I still have trouble with.
1) I wanted to chain multiple Components of 1 particle together (to constriaint X and Y axis for example). But that didn't seem to work. @pabloferz can you take a look? The commented lines in my example are my attempt to do it.
2) If we constraint a particle close to the edge of the box, the simulation becomes unstable and the particle is not constraint.
I have a strong idea, what is going on. Since we use the unwrapped coordinates in HOOMD-blue, once the particle crosses the box boundary, there is a incontinuity in the energy and force. Hence, the particle gets pushed really hard all of a sudden and continues to accelerate, which breaks everything. That would be fixed with #38 .

(There is one minor thing, I noticed. The potential energy associated with biasing is not reported as energy to hoomd-blue, which might be confusing for some users.)

Anyhow, I am marking this as done for now.
Maybe we should open issues for the problems above?

File: pysages/ssages/methods/umbrella.py (hunk 1 above)
Comment (pabloferz) [inline pysages/ssages/methods/umbrella.py:1]:
```suggestion
# SPDX-License-Identifier: MIT
```

**Comment by pabloferz** [comment]
> [hist.pdf](https://github.com/SSAGESLabs/PySAGES/files/7262741/hist.pdf) I am happy, it seems to work at least somehow.
> 
> You can bias the position of a single particle and it fluctuates around a given position. There are two things, I still have trouble with.
> 
>     1. I wanted to chain multiple Components of 1 particle together (to constriaint X and Y axis for example). But that didn't seem to work. @pabloferz can you take a look? The commented lines in my example are my attempt to do it.
> 
>     2. If we constraint a particle close to the edge of the box, the simulation becomes unstable and the particle is not constraint.
>        I have a strong idea, what is going on. Since we use the unwrapped coordinates in HOOMD-blue, once the particle crosses the box boundary, there is a incontinuity in the energy and force. Hence, the particle gets pushed really hard all of a sudden and continues to accelerate, which breaks everything. That would be fixed with [Wrapping coordinates #38](https://github.com/SSAGESLabs/PySAGES/issues/38) .
> 
> 
> (There is one minor thing, I noticed. The potential energy associated with biasing is not reported as energy to hoomd-blue, which might be confusing for some users.)
> 
> Anyhow, I am marking this as done for now. Maybe we should open issues for the problems above?

I'll have a look at 1. But that probably has to be fixed in a separate PR.

For 2, you're right about what the issue is, and as you said it should be fixed by solving #38.

For logging the potential energy taking into account the pysages contribution we would have to do some work on the backends code. I would also open a separate issue for this.

**Comment by pabloferz** [comment]
One thing to note, is that we'd probably also want a separate sampling method that implements umbrella sampling over different regions and allow to compute the weighted histograms and to estimate the free energy. That of course is a task for another PR, but I'm leaving the comment to keep this in mind.

**Comment by InnocentBug** [comment]
Yes, that is a good point.
We should make a new method that implements that.
But it is a bit higher level than this one. And it probably benefits from a task scheduler or something similar, that we are working on.
I think after we work out the framework for it, for example with forward flux, we are good to push this.

File: pysages/ssages/methods/umbrella.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    indices = helpers.indices
+
+    def initialize():
+        bias = np.zeros((natoms, 3))
+        return UmbrellaState(bias, None)
+
+    def update(state, rs, vms, ids):
+        xi, Jxi = cv(rs, indices(ids))
+        D = kspring * (xi - center)
+        bias = (-D * Jxi).reshape(state.bias.shape)
```
Comment (pabloferz) [inline pysages/ssages/methods/umbrella.py:46]:
```suggestion
        bias = (-Jxi.T @ D).reshape(state.bias.shape)
```
does this and passing the following

```python
cvs = (
    Component([0], 0),
    Component([0], 1),
)

center_cv = [[0., 0.]]
```
work?

Comment (InnocentBug) [reply pysages/ssages/methods/umbrella.py:46]: No, unfortunately it does not.
The shapes of the bias seems to mismatch if I try this.

```
  File "/home/ludwig/git/PySAGES/examples/umbrella_sampling/umbrella.py", line 57, in <module>
    main()
  File "/home/ludwig/git/PySAGES/examples/umbrella_sampling/umbrella.py", line 52, in main
    hoomd.run(1e5, limit_multiple=100)
  File "/home/ludwig/git/hoomd-blue/install/hoomd/__init__.py", line 201, in run
    context.current.system.run(int(tsteps), callback_period, callback, limit_hours, int(limit_multiple));
  File "/home/ludwig/.local/lib/python3.9/site-packages/pysages/backends/hoomd.py", line 44, in update
    self.state = self._update(self.snapshot, self.state)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/_src/traceback_util.py", line 162, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/_src/api.py", line 411, in cache_miss
    out_flat = xla.xla_call(
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/core.py", line 1618, in bind
    return call_bind(self, fun, *args, **params)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/core.py", line 1609, in call_bind
    outs = primitive.process(top_trace, fun, tracers, params)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/core.py", line 1621, in process
    return trace.process_call(self, fun, tracers, params)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/core.py", line 615, in process_call
    return primitive.impl(f, *tracers, **params)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/interpreters/xla.py", line 622, in _xla_call_impl
    compiled_fun = _xla_callable(fun, device, backend, name, donated_invars,
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/linear_util.py", line 262, in memoized_fun
    ans = call(fun, *args)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/interpreters/xla.py", line 694, in _xla_callable
    return lower_xla_callable(fun, device, backend, name, donated_invars, *arg_specs).compile().unsafe_call
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/interpreters/xla.py", line 702, in lower_xla_callable
    jaxpr, out_avals, consts = pe.trace_to_jaxpr_final(
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/interpreters/partial_eval.py", line 1522, in trace_to_jaxpr_final
    jaxpr, out_avals, consts = trace_to_subjaxpr_dynamic(fun, main, in_avals)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/interpreters/partial_eval.py", line 1500, in trace_to_subjaxpr_dynamic
    ans = fun.call_wrapped(*in_tracers)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/linear_util.py", line 166, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "/home/ludwig/.local/lib/python3.9/site-packages/pysages/ssages/methods/core.py", line 74, in update
    return _update(state, rs, vms, ids)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/_src/traceback_util.py", line 162, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/_src/api.py", line 411, in cache_miss
    out_flat = xla.xla_call(
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/core.py", line 1618, in bind
    return call_bind(self, fun, *args, **params)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/core.py", line 1609, in call_bind
    outs = primitive.process(top_trace, fun, tracers, params)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/core.py", line 1621, in process
    return trace.process_call(self, fun, tracers, params)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/interpreters/partial_eval.py", line 1310, in process_call
    jaxpr, out_avals, consts = trace_to_subjaxpr_dynamic(f, self.main, in_avals)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/interpreters/partial_eval.py", line 1500, in trace_to_subjaxpr_dynamic
    ans = fun.call_wrapped(*in_tracers)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/linear_util.py", line 166, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "/home/ludwig/.local/lib/python3.9/site-packages/pysages/ssages/methods/umbrella.py", line 46, in update
    bias = (-D * Jxi).reshape(state.bias.shape)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/core.py", line 515, in __mul__
    def __mul__(self, other): return self.aval._mul(self, other)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/_src/numpy/lax_numpy.py", line 6553, in deferring_binary_op
    return binary_op(self, other)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/_src/traceback_util.py", line 162, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/_src/api.py", line 411, in cache_miss
    out_flat = xla.xla_call(
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/core.py", line 1618, in bind
    return call_bind(self, fun, *args, **params)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/core.py", line 1609, in call_bind
    outs = primitive.process(top_trace, fun, tracers, params)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/core.py", line 1621, in process
    return trace.process_call(self, fun, tracers, params)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/interpreters/partial_eval.py", line 1310, in process_call
    jaxpr, out_avals, consts = trace_to_subjaxpr_dynamic(f, self.main, in_avals)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/interpreters/partial_eval.py", line 1500, in trace_to_subjaxpr_dynamic
    ans = fun.call_wrapped(*in_tracers)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/linear_util.py", line 166, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/_src/numpy/lax_numpy.py", line 686, in fn
    return lax_fn(x1, x2) if x1.dtype != bool_ else bool_lax_fn(x1, x2)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/_src/lax/lax.py", line 354, in mul
    return mul_p.bind(x, y)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/core.py", line 267, in bind
    out = top_trace.process_primitive(self, tracers, params)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/interpreters/partial_eval.py", line 1297, in process_primitive
    out_avals = primitive.abstract_eval(*avals, **params)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/_src/lax/lax.py", line 2151, in standard_abstract_eval
    return ShapedArray(shape_rule(*avals, **kwargs), dtype_rule(*avals, **kwargs),
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/_src/lax/lax.py", line 2247, in _broadcasting_shape_rule
    raise TypeError(msg.format(name, ', '.join(map(str, map(tuple, shapes)))))
jax._src.traceback_util.UnfilteredStackTrace: TypeError: mul got incompatible shapes for broadcasting: (1, 2), (2, 600).

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/ludwig/git/PySAGES/examples/umbrella_sampling/umbrella.py", line 57, in <module>
    main()
  File "/home/ludwig/git/PySAGES/examples/umbrella_sampling/umbrella.py", line 52, in main
    hoomd.run(1e5, limit_multiple=100)
  File "/home/ludwig/git/hoomd-blue/install/hoomd/__init__.py", line 201, in run
    context.current.system.run(int(tsteps), callback_period, callback, limit_hours, int(limit_multiple));
  File "/home/ludwig/.local/lib/python3.9/site-packages/pysages/backends/hoomd.py", line 44, in update
    self.state = self._update(self.snapshot, self.state)
  File "/home/ludwig/.local/lib/python3.9/site-packages/pysages/ssages/methods/core.py", line 74, in update
    return _update(state, rs, vms, ids)
  File "/home/ludwig/.local/lib/python3.9/site-packages/pysages/ssages/methods/umbrella.py", line 46, in update
    bias = (-D * Jxi).reshape(state.bias.shape)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/_src/numpy/lax_numpy.py", line 6553, in deferring_binary_op
    return binary_op(self, other)
  File "/home/ludwig/.local/lib/python3.9/site-packages/jax/_src/numpy/lax_numpy.py", line 686, in fn
    return lax_fn(x1, x2) if x1.dtype != bool_ else bool_lax_fn(x1, x2)
TypeError: mul got incompatible shapes for broadcasting: (1, 2), (2, 600).

Comment (InnocentBug) [reply pysages/ssages/methods/umbrella.py:46]: I tried debugging it a little bit by hand, but couldn't find a handle of how to deal with it.

Comment (pabloferz) [reply pysages/ssages/methods/umbrella.py:46]: You're stacktrace is showing `bias = (-D * Jxi).reshape(state.bias.shape)`. Did you checked with: `bias = (-Jxi.T @ D).reshape(state.bias.shape)`

Comment (InnocentBug) [reply pysages/ssages/methods/umbrella.py:46]: it works for the one 1D case, but not with more CVs
but I think you are on the right path

**Comment by InnocentBug** [comment]
Just added a test in the python script.
But I haven't added it as CI, since we haven't decided how we do tests that require one of the backend simulations.
We could use the docker container. Or we use a fixed installation env. like on midway3.

File: pysages/ssages/methods/umbrella.py (hunk 4 above)
Comment (pabloferz) [reply pysages/ssages/methods/umbrella.py:46]: Do you have a stacktrace for the 2D case?

**Comment by pabloferz** [comment]
I'd be inclined to test right now with the docker until we discuss with RCC our midway3-based CI options, if any.

File: pysages/ssages/methods/umbrella.py (hunk 4 above)
Comment (InnocentBug) [reply pysages/ssages/methods/umbrella.py:46]:
```
Traceback (most recent call last):
  File "umbrella.py", line 51, in <module>
    main()
  File "umbrella.py", line 46, in main
    hoomd.run(1e5, limit_multiple=100)
  File "/software/python-anaconda-2020.11-el8-x86_64/envs/pysages/lib/python3.8/site-packages/hoomd/__init__.py", line 201, in run
    context.current.system.run(int(tsteps), callback_period, callback, limit_hours, int(limit_multiple));
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/pysages/backends/hoomd.py", line 44, in update
    self.state = self._update(self.snapshot, self.state)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/_src/traceback_util.py", line 183, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/_src/api.py", line 399, in cache_miss
    out_flat = xla.xla_call(
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/core.py", line 1561, in bind
    return call_bind(self, fun, *args, **params)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/core.py", line 1552, in call_bind
    outs = primitive.process(top_trace, fun, tracers, params)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/core.py", line 1564, in process
    return trace.process_call(self, fun, tracers, params)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/core.py", line 607, in process_call
    return primitive.impl(f, *tracers, **params)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/interpreters/xla.py", line 607, in _xla_call_impl
    compiled_fun = _xla_callable(fun, device, backend, name, donated_invars,
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/linear_util.py", line 262, in memoized_fun
    ans = call(fun, *args)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/interpreters/xla.py", line 682, in _xla_callable
    jaxpr, out_avals, consts = pe.trace_to_jaxpr_final(
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/interpreters/partial_eval.py", line 1285, in trace_to_jaxpr_final
    jaxpr, out_avals, consts = trace_to_subjaxpr_dynamic(fun, main, in_avals)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/interpreters/partial_eval.py", line 1263, in trace_to_subjaxpr_dynamic
    ans = fun.call_wrapped(*in_tracers)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/linear_util.py", line 166, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/pysages/ssages/methods/core.py", line 74, in update
    return _update(state, rs, vms, ids)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/_src/traceback_util.py", line 183, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/_src/api.py", line 399, in cache_miss
    out_flat = xla.xla_call(
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/core.py", line 1561, in bind
    return call_bind(self, fun, *args, **params)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/core.py", line 1552, in call_bind
    outs = primitive.process(top_trace, fun, tracers, params)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/core.py", line 1564, in process
    return trace.process_call(self, fun, tracers, params)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/interpreters/partial_eval.py", line 1072, in process_call
    jaxpr, out_avals, consts = trace_to_subjaxpr_dynamic(f, self.main, in_avals)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/interpreters/partial_eval.py", line 1263, in trace_to_subjaxpr_dynamic
    ans = fun.call_wrapped(*in_tracers)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/linear_util.py", line 166, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/pysages/ssages/methods/umbrella.py", line 46, in update
    bias = (-D * Jxi).reshape(state.bias.shape)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/core.py", line 507, in __mul__
    def __mul__(self, other): return self.aval._mul(self, other)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 5871, in deferring_binary_op
    return binary_op(self, other)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 431, in fn
    return lax_fn(x1, x2) if x1.dtype != bool_ else bool_lax_fn(x1, x2)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/_src/lax/lax.py", line 348, in mul
    return mul_p.bind(x, y)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/core.py", line 264, in bind
    out = top_trace.process_primitive(self, tracers, params)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/interpreters/partial_eval.py", line 1059, in process_primitive
    out_avals = primitive.abstract_eval(*avals, **params)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/_src/lax/lax.py", line 2136, in standard_abstract_eval
    return ShapedArray(shape_rule(*avals, **kwargs), dtype_rule(*avals, **kwargs),
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/_src/lax/lax.py", line 2232, in _broadcasting_shape_rule
    raise TypeError(msg.format(name, ', '.join(map(str, map(tuple, shapes)))))
jax._src.traceback_util.UnfilteredStackTrace: TypeError: mul got incompatible shapes for broadcasting: (1, 3), (3, 600).

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "umbrella.py", line 51, in <module>
    main()
  File "umbrella.py", line 46, in main
    hoomd.run(1e5, limit_multiple=100)
  File "/software/python-anaconda-2020.11-el8-x86_64/envs/pysages/lib/python3.8/site-packages/hoomd/__init__.py", line 201, in run
    context.current.system.run(int(tsteps), callback_period, callback, limit_hours, int(limit_multiple));
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/pysages/backends/hoomd.py", line 44, in update
    self.state = self._update(self.snapshot, self.state)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/pysages/ssages/methods/core.py", line 74, in update
    return _update(state, rs, vms, ids)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/pysages/ssages/methods/umbrella.py", line 46, in update
    bias = (-D * Jxi).reshape(state.bias.shape)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 5871, in deferring_binary_op
    return binary_op(self, other)
  File "/home/ludwigschneider/.local/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 431, in fn
    return lax_fn(x1, x2) if x1.dtype != bool_ else bool_lax_fn(x1, x2)
TypeError: mul got incompatible shapes for broadcasting: (1, 3), (3, 600).
[midway3-login1.rcc.local:mpi_rank_0][deallocate_vbuf_region] [src/mpid/ch3/channels/mrail/src/gen2/vbuf.c:238] cuda failed with 3 (initialization error)

**Comment by InnocentBug** [comment]
> I'd be inclined to test right now with the docker until we discuss with RCC our midway3-based CI options, if any.

The best way to do this, would be to host a "base" docker container on docker hub.
In the test env we pull this docker container from there and install the fresh pysages into it. and then run the tests there.

Freshly building the docker container from scratch takes definitely too long.
The only downside is, that any time we change the base container, we have to separately write the docker file and push the new container to docker hub.

Any ideas comments to this approach, before I start implementing it?

**Comment by pabloferz** [comment]
> The only downside is, that any time we change the base container, we have to separately write the docker file and push the new container to docker hub.

I do not anticipate changing the OpenMM and HOOMD-blue installations that often. Is it possible to have it so we can rebuild `hoomd-dlext` or `openmm-dlext` on a built container?

**Comment by InnocentBug** [comment]
I can stack as many container as I want.

So make a base container without the plugins. Build the next layer of container with the plugins.
And finally pull this container, to build pysages into it.

Only downside is, that we cannot separate both plugins separately.
But build time shouldn't be too bad for this.

File: pysages/ssages/methods/umbrella.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    def __init__(self, cvs, kspring, center, *args, **kwargs):
+        super().__init__(cvs, args, kwargs)
+        self.kspring = np.asarray(kspring)
+        self.center = np.asarray(center)
+
+    def __call__(self, snapshot, helpers):
+        return _umbrella(snapshot, self.cv, self.center, self.kspring, helpers)
+
+
+def _umbrella(snapshot, cv, center, kspring, helpers):
```
Comment (pabloferz) [inline pysages/ssages/methods/umbrella.py:35]:
```suggestion
    def __call__(self, snapshot, helpers):
        return _umbrella(self, snapshot, helpers)


def _umbrella(method, snapshot, helpers):
    cv = method.cv
    center = method.center.reshape(1, -1)
    kspring = self.kspring.reshape(1, -1)
```

Does this (along with the previous matmul change) fix the issue in higher dims than one?

**Comment by InnocentBug** [comment]
No, unfortunately it doesn't solve the problem.
I just liked it as being more general.

Oct 4, 2021 18:59:59 Pablo Zubieta ***@***.***>:

**Comment by InnocentBug** [comment]
Wait, sorry. I am on my phone and missed that you suggested a change. I try it tomorrow

Oct 4, 2021 18:59:59 Pablo Zubieta ***@***.***>:

File: pysages/ssages/methods/umbrella.py (hunk 5 above)
Comment (InnocentBug) [reply pysages/ssages/methods/umbrella.py]: Ok, I played around with this and actually used my brain.
Now it works, but there is probably a faster way to write that down.
We need to sum over the different CVs.

File: examples/umbrella_sampling/.gitignore [hunk 6]
```diff
@@ -0,0 +1,4 @@
+start.gsd
+umbrella.gsd
+umbrella.dat
+hist.pdf
```
Comment (pabloferz) [inline examples/umbrella_sampling/.gitignore:4]: Could this be placed in the main gitignore as `examples/*.gsd`, `examples/*.dat`, `examples/*.pdf`?

**Review by pabloferz — CHANGES_REQUESTED** [review]

File: examples/umbrella_sampling/.gitignore (hunk 6 above)
Comment (InnocentBug) [reply examples/umbrella_sampling/.gitignore]: sure, I put them in the global .gitignore noe.
And have it ignore all sub directories of example.

**Review by pabloferz — APPROVED** [review]
I think this is ready. I'll go ahead and Squash and merge.

**Pull request merged**